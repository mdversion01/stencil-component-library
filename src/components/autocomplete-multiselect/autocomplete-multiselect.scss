.ac-multi-select-container {
  border: 1px solid #ced4da;
  border-radius: 0.25rem;
  display: flex;
  flex-wrap: wrap;
  width: 100%;

  .input-group {
    flex-wrap: nowrap !important;

    .form-control {
      border: none;
      display: flex !important;
      flex: 1 1 auto !important; // 200px
      padding: 0.375rem 0.75rem;
      width: 100%;
    }

    > .input-group-btn {
      border: none;
    }
  }

  &:invalid,
  &.is-invalid {
    border-color: #b30009;
  }

  .form-control:focus {
    background-color: transparent;
    border-color: none;
    color: #383838;
    outline: 0;
    box-shadow: none;
  }

  &.ac-focused {
    border-color: #9bc4f6 !important;
    box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25) !important;
  }

  &.is-invalid-focused {
    border-color: #dc3545 !important; /* Fallback for --bs-form-invalid-border-color */
    box-shadow: 0 0 0 0.25rem rgba(220, 53, 69, 0.25) !important; /* Fallback for --bs-danger-rgb */
  }

  &.sm {
    min-height: 36px;

    .input-group-btn {
      min-height: 32px;
    }
  }

  &.lg {
    min-height: 48px;
  }

  .input-sm {
    min-height: 34px;
  }

  .input-lg {
    min-height: 46px;
  }
}

.ac-selected-items {
  display: flex;
  flex-wrap: wrap;
  overflow: auto;
  max-height: 100px;


  .badge {
    align-items: center;
    background-color: transparent;
    border: 1px solid #8e8e8e;
    // border-radius: 1rem;
    color: #000000;
    display: flex;
    font-size: 0.75rem;
    height: 30px;
    justify-content: center;
    margin: 0.15rem;
    padding: 0 0.5rem;
    position: relative;

    &.sm {
      font-size: 0.65rem !important;
      height: 28px;
    }

    &.lg {
      font-size: 1rem !important;
      height: 42px;
    }

    &.square {
      border-radius: 0px;
    }

    &.pill {
      border-radius: 1.75rem;
      padding: 0.25rem 0.5rem;
    }

    span {
      margin-right: 0.5rem;
      margin-top: -0.1rem;
    }

    .remove-btn {
      background-color: transparent;
      border: none;
      border-radius: 25%;
      margin: 0;
      padding: 0;
      height: 1rem;
      width: 1rem;
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .remove-btn svg {
      cursor: pointer;
      fill: #666666;
      height: 1rem;
      position: relative;
      width: 1rem;
      z-index: 1;
    }

    .remove-btn:hover svg,
    .remove-btn:focus svg {
      fill: #b30009;
      outline: none;
    }
  }

  .text-bg-primary {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #f7caca;
        }
      }
    }
  }

  .text-bg-secondary {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #f7caca;
        }
      }
    }
  }

  .text-bg-success {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #f7caca;
        }
      }
    }
  }

  .text-bg-danger {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #f7caca;
        }
      }
    }
  }

  .text-bg-warning {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #a90000;
        }
      }
    }
  }

  .text-bg-info {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #a90000;
        }
      }
    }
  }

  .text-bg-dark {
    .remove-btn {
      svg {
        fill: #fff;
        &:hover,
        &focus {
          fill: #f7caca;
        }
      }
    }
  }
}
