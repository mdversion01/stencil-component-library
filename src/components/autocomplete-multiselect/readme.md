# autocomplete-multiselect



<!-- Auto Generated Below -->


## Properties

| Property                  | Attribute                     | Description                                                                            | Type                             | Default                      |
| ------------------------- | ----------------------------- | -------------------------------------------------------------------------------------- | -------------------------------- | ---------------------------- |
| `addBtn`                  | `add-btn`                     |                                                                                        | `boolean`                        | `false`                      |
| `addIcon`                 | `add-icon`                    |                                                                                        | `string`                         | `''`                         |
| `addNewOnEnter`           | `add-new-on-enter`            |                                                                                        | `boolean`                        | `true`                       |
| `arialabelledBy`          | `arialabelled-by`             |                                                                                        | `string`                         | `''`                         |
| `autoSort`                | `auto-sort`                   |                                                                                        | `boolean`                        | `true`                       |
| `badgeInlineStyles`       | `badge-inline-styles`         |                                                                                        | `string`                         | `''`                         |
| `badgeShape`              | `badge-shape`                 |                                                                                        | `string`                         | `''`                         |
| `badgeVariant`            | `badge-variant`               |                                                                                        | `string`                         | `''`                         |
| `clearIcon`               | `clear-icon`                  |                                                                                        | `string`                         | `''`                         |
| `clearInputOnBlurOutside` | `clear-input-on-blur-outside` |                                                                                        | `boolean`                        | `false`                      |
| `devMode`                 | `dev-mode`                    |                                                                                        | `boolean`                        | `false`                      |
| `disabled`                | `disabled`                    |                                                                                        | `boolean`                        | `false`                      |
| `editable`                | `editable`                    |                                                                                        | `boolean`                        | `false`                      |
| `error`                   | `error`                       |                                                                                        | `boolean`                        | `false`                      |
| `errorMessage`            | `error-message`               |                                                                                        | `string`                         | `''`                         |
| `formId`                  | `form-id`                     |                                                                                        | `string`                         | `''`                         |
| `formLayout`              | `form-layout`                 |                                                                                        | `"" \| "horizontal" \| "inline"` | `''`                         |
| `inputCol`                | `input-col`                   |                                                                                        | `number`                         | `10`                         |
| `inputCols`               | `input-cols`                  |                                                                                        | `string`                         | `''`                         |
| `inputId`                 | `input-id`                    |                                                                                        | `string`                         | `''`                         |
| `label`                   | `label`                       |                                                                                        | `string`                         | `''`                         |
| `labelCol`                | `label-col`                   |                                                                                        | `number`                         | `2`                          |
| `labelCols`               | `label-cols`                  | Responsive column class specs (e.g., "col", "col-sm-3 col-md-4", or "xs-12 sm-6 md-4") | `string`                         | `''`                         |
| `labelHidden`             | `label-hidden`                |                                                                                        | `boolean`                        | `false`                      |
| `labelSize`               | `label-size`                  |                                                                                        | `"" \| "lg" \| "sm"`             | `''`                         |
| `name`                    | `name`                        |                                                                                        | `string`                         | `undefined`                  |
| `options`                 | `options`                     |                                                                                        | `string[]`                       | `[]`                         |
| `placeholder`             | `placeholder`                 |                                                                                        | `string`                         | `'Type to search/filter...'` |
| `preserveInputOnSelect`   | `preserve-input-on-select`    |                                                                                        | `boolean`                        | `undefined`                  |
| `rawInputName`            | `raw-input-name`              |                                                                                        | `string`                         | `undefined`                  |
| `removeClearBtn`          | `remove-clear-btn`            |                                                                                        | `boolean`                        | `false`                      |
| `required`                | `required`                    |                                                                                        | `boolean`                        | `false`                      |
| `size`                    | `size`                        |                                                                                        | `"" \| "lg" \| "sm"`             | `''`                         |
| `type`                    | `type`                        |                                                                                        | `string`                         | `''`                         |
| `validation`              | `validation`                  |                                                                                        | `boolean`                        | `false`                      |
| `validationMessage`       | `validation-message`          |                                                                                        | `string`                         | `''`                         |


## Events

| Event               | Description                                 | Type                                                                                          |
| ------------------- | ------------------------------------------- | --------------------------------------------------------------------------------------------- |
| `clear`             |                                             | `CustomEvent<void>`                                                                           |
| `componentError`    |                                             | `CustomEvent<{ message: string; stack?: string; }>`                                           |
| `itemSelect`        |                                             | `CustomEvent<string>`                                                                         |
| `multiSelectChange` |                                             | `CustomEvent<string[]>`                                                                       |
| `optionDelete`      |                                             | `CustomEvent<string>`                                                                         |
| `optionsChange`     | ðŸ”” Hook for hosts to mirror/persist options | `CustomEvent<{ options: string[]; reason: "replace" \| "add" \| "delete"; value?: string; }>` |


## Methods

### `filterOptions() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `getOptions() => Promise<string[]>`

ðŸ”Ž Read current options from the component (for hosts).

#### Returns

Type: `Promise<string[]>`



### `navigateOptions(direction: number) => Promise<void>`



#### Parameters

| Name        | Type     | Description |
| ----------- | -------- | ----------- |
| `direction` | `number` |             |

#### Returns

Type: `Promise<void>`



### `setOptions(next: string[]) => Promise<void>`

ðŸ”§ Replace options from the host (for hosts). Also emits optionsChange('replace').

#### Parameters

| Name   | Type       | Description |
| ------ | ---------- | ----------- |
| `next` | `string[]` |             |

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
