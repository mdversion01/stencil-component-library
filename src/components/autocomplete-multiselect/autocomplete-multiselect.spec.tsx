// src/components/autocomplete-multiselect/autocomplete-multiselect.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { AutocompleteMultiselect } from './autocomplete-multiselect';
import { describe, it, expect, jest } from '@jest/globals';

describe('autocomplete-multiselect', () => {
  it('renders component', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    expect(page.root).toBeTruthy();
  });

  it('opens dropdown only when there are matches; stays closed when no input', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect input-id="fruit" label="Fruits"></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Apple', 'Banana', 'Orange'];

    // No input -> closed
    comp.inputValue = '';
    await comp.filterOptions();
    await page.waitForChanges();

    expect(comp.filteredOptions).toEqual([]);
    expect((comp as any)['dropdownOpen']).toBe(false);
    const input = page.root!.querySelector('input')!;
    expect(input.getAttribute('aria-expanded')).toBe('false');
    expect(page.root!.querySelector('.autocomplete-dropdown')).toBeNull();

    // With matches -> open
    comp.inputValue = 'an'; // Banana, Orange
    await comp.filterOptions();
    await page.waitForChanges();

    expect(comp.filteredOptions).toEqual(['Banana', 'Orange']);
    expect((comp as any)['dropdownOpen']).toBe(true);
    expect(input.getAttribute('aria-expanded')).toBe('true');
    expect(page.root!.querySelector('.autocomplete-dropdown')).toBeTruthy();
  });

  it('stays closed when there is input but no matches', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect input-id="fruit" label="Fruits"></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Apple', 'Banana', 'Orange'];

    comp.inputValue = 'zzz';
    await comp.filterOptions();
    await page.waitForChanges();

    expect(comp.filteredOptions).toEqual([]);
    expect((comp as any)['dropdownOpen']).toBe(false);
    const input = page.root!.querySelector('input')!;
    expect(input.getAttribute('aria-expanded')).toBe('false');
    expect(page.root!.querySelector('.autocomplete-dropdown')).toBeNull();
  });

  it('filters options based on input (case-insensitive)', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Apple', 'Banana', 'Orange'];
    comp.inputValue = 'ap';
    await comp.filterOptions();
    await page.waitForChanges();
    expect(comp.filteredOptions).toEqual(['Apple']);
  });

  it('selects and toggles item correctly', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['One', 'Two'];

    comp.toggleItem('One');
    await page.waitForChanges();
    expect(comp.selectedItems).toContain('One');

    comp.toggleItem('One');
    await page.waitForChanges();
    expect(comp.selectedItems).not.toContain('One');
  });

  it('validate returns false when required and nothing selected', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect required></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    const valid = comp.validate();
    await page.waitForChanges();
    expect(valid).toBe(false);
    expect(comp.validation).toBe(true);
  });

  it('validate returns true when required and item selected', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect required></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.selectedItems = ['One'];
    comp.inputValue = 'dummy';
    await page.waitForChanges();
    const valid = comp.validate();
    expect(valid).toBe(true);
    expect(comp.validation).toBe(false);
  });

  // Enter adds brand-new value (editable=true) and selects it; autoSort default true
  it('adds brand-new value on Enter, upserts into options and selects it (sorted by default)', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect editable></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Banana', 'Orange'];
    comp.inputValue = 'Apple';
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    expect(comp.options).toEqual(['Apple', 'Banana', 'Orange']);
    expect(comp.selectedItems).toContain('Apple');
    expect(comp.inputValue).toBe('');
  });

  // auto-sort disabled preserves insertion order
  it('preserves insertion order when auto-sort is false', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect editable auto-sort="false"></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Orange', 'Banana'];
    comp.inputValue = 'Apple';
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    expect(comp.options).toEqual(['Orange', 'Banana', 'Apple']);
    expect(comp.selectedItems).toContain('Apple');
  });

  // does not add duplicate (case-insensitive)
  it('does not duplicate existing options (case-insensitive) when pressing Enter', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Apple', 'Banana'];
    comp.inputValue = 'Apple';
    await comp.filterOptions(); // ensure filtered includes 'Apple'
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    expect(comp.options).toEqual(['Apple', 'Banana']);
    expect(comp.selectedItems).toContain('Apple');
  });

  // add-new-on-enter can be disabled
  it('respects add-new-on-enter="false" (does not upsert new values on Enter)', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect editable add-new-on-enter="false"></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Banana', 'Orange'];
    comp.inputValue = 'Kiwi';
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    expect(comp.options).toEqual(['Banana', 'Orange']);
    expect(comp.selectedItems).not.toContain('Kiwi');
  });

  // addBtn visibility: shown only when editable && addBtn && input has text
  it('shows addBtn only when input has text (editable & add-btn)', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect add-btn editable></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;

    // empty -> hidden
    comp.inputValue = '';
    await page.waitForChanges();
    expect(page.root!.querySelector('button.add-btn')).toBeNull();

    // text -> visible
    comp.inputValue = 'Cherry';
    await page.waitForChanges();
    expect(page.root!.querySelector('button.add-btn')).toBeTruthy();

    // disabled -> still rendered but disabled
    page.root!.setAttribute('disabled', '');
    await page.waitForChanges();
    const btn = page.root!.querySelector('button.add-btn')!;
    expect(btn).toHaveAttribute('disabled');
  });

  // add button upserts + selects (and emits itemSelect)
  it('emits itemSelect and upserts + selects on addBtn click', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect add-btn editable></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Banana', 'Orange'];

    const spy = jest.fn();
    page.root!.addEventListener('itemSelect', spy);

    comp.inputValue = 'Cherry';
    await page.waitForChanges();

    const btn = page.root!.querySelector('button.add-btn') as HTMLButtonElement;
    btn.click();
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledWith(expect.objectContaining({ detail: 'Cherry' }));
    expect(comp.options).toEqual(['Banana', 'Cherry', 'Orange']);
    expect(comp.selectedItems).toContain('Cherry');
    expect(comp.inputValue).toBe('');
  });

  it('handles arrow key navigation and enter selection', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['One', 'Two', 'Three'];
    comp.inputValue = 'T';
    await comp.filterOptions();
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    // Enter should pick the first filtered option ('Two')
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    expect(comp.selectedItems).toContain('Two');
    expect(comp.inputValue).toBe('');
  });

  it('aria-activedescendant updates on focus change', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect options='["A", "B"]' input-id="letters"></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.inputValue = 'B';
    await comp.filterOptions();
    comp.focusedOptionIndex = 0;
    await page.waitForChanges();
    const input = page.root!.querySelector('input')!;
    expect(input.getAttribute('aria-activedescendant')).toContain('letters-option-0');
  });

  it('scrolls focused item into view', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;

    comp.options = ['One', 'Two', 'Three'];
    comp.inputValue = 'o';
    await comp.filterOptions();
    comp.focusedOptionIndex = -1;
    await page.waitForChanges();

    const originalQS = HTMLElement.prototype.querySelectorAll;
    const fakeItems = ['One', 'Two', 'Three'].map(() => {
      const li = document.createElement('li') as any;
      li.scrollIntoView = jest.fn();
      return li;
    });
    (HTMLElement.prototype.querySelectorAll as any) = jest.fn(() => fakeItems as any);

    try {
      jest.useFakeTimers();
      await comp.navigateOptions(1);
      jest.runAllTimers();
      expect(fakeItems[0].scrollIntoView as any).toHaveBeenCalled();
    } finally {
      HTMLElement.prototype.querySelectorAll = originalQS as any;
      jest.useRealTimers();
    }
  });

  it('renders badges with class and styles', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect badge-variant="success" badge-shape="rounded" badge-inline-styles="color: white;"></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.selectedItems = ['Item'];
    await page.waitForChanges();

    const badge = page.root!.querySelector('.badge')!;
    expect(badge.classList.contains('text-bg-success')).toBe(true);
    expect(badge.classList.contains('rounded')).toBe(true);
    expect(badge.getAttribute('style')!).toContain('color: white');
  });

  it('emits multiSelectChange on selection', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });

    const spy = jest.fn();
    page.root!.addEventListener('multiSelectChange', spy);
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['A'];
    comp.toggleItem('A');
    await page.waitForChanges();

    expect(spy).toHaveBeenCalled();
  });

  it('clears input and items with clearAll()', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.selectedItems = ['A'];
    comp.inputValue = 'text';
    comp.clearAll();
    await page.waitForChanges();

    expect(comp.selectedItems.length).toBe(0);
    expect(comp.inputValue).toBe('');
  });

  it('closes dropdown when clicking outside (closeDropdown path)', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;

    // Open the dropdown
    comp.options = ['a', 'b'];
    comp.inputValue = 'a';
    await comp.filterOptions();
    await page.waitForChanges();
    expect((comp as any)['dropdownOpen']).toBe(true);

    // Simulate a click outside the component
    const fakeEvent = {
      composedPath: () => [document.body],
    } as unknown as MouseEvent;

    (comp as any).handleClickOutside(fakeEvent);
    await page.waitForChanges();

    expect(comp.isFocused).toBe(false);
    expect((comp as any)['dropdownOpen']).toBe(false);
  });

  it('disables input and add & clear buttons when disabled is true', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect disabled add-btn editable></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;

    comp.inputValue = 'foo';
    await page.waitForChanges();

    const input = page.root!.querySelector('input')!;
    const addBtn = page.root!.querySelector('button.add-btn')!;
    const clearBtn = page.root!.querySelector('button.clear-btn')!;

    expect(input).toHaveAttribute('disabled');
    expect(addBtn).toHaveAttribute('disabled');
    expect(clearBtn).toHaveAttribute('disabled');
  });

  it('sets correct ARIA attributes when required and invalid', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect required validation-message="Required field."></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.validate();
    await page.waitForChanges();
    const input = page.root!.querySelector('input')!;
    expect(input.getAttribute('aria-required')).toBe('true');
    const group = page.root!.querySelector('.input-group')!;
    expect(group.classList.contains('is-invalid')).toBe(true);
  });

  it('emits clear event when cleared', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    const clearSpy = jest.fn();
    page.root!.addEventListener('clear', clearSpy);

    comp.selectedItems = ['A'];
    comp.inputValue = 'test';
    comp.clearAll();
    await page.waitForChanges();

    expect(clearSpy).toHaveBeenCalled();
  });

  it('renders error message when error is true', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect error error-message="Something went wrong"></autocomplete-multiselect>`,
    });
    await page.waitForChanges();
    const msg = page.root!.querySelector('.error-message')!;
    expect(msg.textContent).toContain('Something went wrong');
  });

  it('parses kebab-case badge styles to camelCase', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect badge-inline-styles="font-size: 14px;"></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.selectedItems = ['Styled'];
    await page.waitForChanges();
    const badge = page.root!.querySelector('.badge')!;
    expect(badge.getAttribute('style')!).toContain('font-size: 14px');
  });

  it('uses inputId and label for accessibility linkage', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect input-id="fruit-input" label="Fruits" arialabelled-by="label-id"></autocomplete-multiselect>`,
    });

    const input = page.root!.querySelector('input')!;
    expect(input.getAttribute('aria-controls')).toContain('fruitInput-listbox');
    expect(input.getAttribute('aria-labelledby')).toBe('label-id');
  });

  // ---- New: optionsChange hook + setOptions/getOptions API ----

  it('emits optionsChange on add (reason="add") and delete (reason="delete")', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect editable></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['Banana'];

    const events: any[] = [];
    page.root!.addEventListener('optionsChange', (e: any) => events.push(e.detail));

    // Add via Enter
    comp.inputValue = 'Apple';
    const input = page.root!.querySelector('input')!;
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true, composed: true }));
    await page.waitForChanges();

    // Delete via internal method (simpler than UI path in unit test)
    (comp as any).deleteUserOption('Apple');
    await page.waitForChanges();

    // Validate events
    const addEvt = events.find(e => e?.reason === 'add');
    const delEvt = events.find(e => e?.reason === 'delete');

    expect(addEvt).toBeTruthy();
    expect(addEvt.value).toBe('Apple');
    expect(Array.isArray(addEvt.options)).toBe(true);

    expect(delEvt).toBeTruthy();
    expect(delEvt.value).toBe('Apple');
    expect(Array.isArray(delEvt.options)).toBe(true);
  });

  it('setOptions replaces options and emits optionsChange with reason="replace"; getOptions returns a copy', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect></autocomplete-multiselect>`,
    });
    const comp = page.rootInstance as AutocompleteMultiselect;

    const spy = jest.fn();
    page.root!.addEventListener('optionsChange', spy);

    await comp.setOptions(['X', 'Y']);
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: expect.objectContaining({
          reason: 'replace',
          options: ['X', 'Y'],
        }),
      }),
    );

    const got = await comp.getOptions();
    expect(got).toEqual(['X', 'Y']);

    // ensure it’s a copy
    got.push('Z');
    const got2 = await comp.getOptions();
    expect(got2).toEqual(['X', 'Y']);
  });

  it('renders full DOM output snapshot for given props', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultiselect],
      html: `<autocomplete-multiselect
               badge-variant="danger"
               required
               label="Tags"
               input-id="tag-input"
               badge-shape="rounded"
               badge-inline-styles="background-color: red; color: white;"
               error
               error-message="Invalid input"
               validation-message="Required field."
               add-btn
               editable
             ></autocomplete-multiselect>`,
    });

    const comp = page.rootInstance as AutocompleteMultiselect;
    comp.options = ['One', 'Two'];
    comp.selectedItems = ['Two'];
    comp.inputValue = 'T';
    await comp.filterOptions();
    comp.validation = true;
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });
});
