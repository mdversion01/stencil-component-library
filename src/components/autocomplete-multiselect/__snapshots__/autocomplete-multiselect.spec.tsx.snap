// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete-multiselect renders full DOM output snapshot for given props 1`] = `
<autocomplete-multiselect add-btn="" badge-inline-styles="background-color: red; color: white;" badge-shape="rounded" badge-variant="danger" editable="" error="" error-message="Invalid input" input-id="tag-input" label="Tags" required="" validation-message="Required field.">
  <div class="autocomplete-container form-group">
    <div>
      <label class="form-control-label invalid" htmlfor="tagInput">
        Tags
      </label>
      <div>
        <div class="ac-multi-select-container is-invalid">
          <div class="ac-selected-items">
            <div class="badge rounded text-bg-danger" style="background-color: red; color: white;">
              <span>
                Two
              </span>
              <button aria-label="Remove Two" class="remove-btn" data-tag="Two" role="button" title="Remove Tag">
                <svg viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9z"></path>
                </svg>
              </button>
            </div>
          </div>
          <div class="ac-input-container">
            <div class="input-group is-invalid">
              <input aria-activedescendant="tagInput-option-0" aria-autocomplete="list" aria-controls="tagInput-listbox" aria-describedby="tagInput-validation" aria-expanded="false" aria-haspopup="listbox" aria-labelledby="" aria-required="true" autocomplete="off" class="form-control" id="tagInput" inputmode="text" placeholder="Tags" role="combobox" spellcheck="false" type="" value="T">
              <button aria-label="Add item" class="add add-btn input-group-btn" role="button" title="Add item" type="button">
                <svg viewBox="0 0 448 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"></path>
                </svg>
              </button>
              <button aria-label="Clear input" class="clear clear-btn input-group-btn" title="Clear input" type="button">
                <svg class="fa-icon" viewBox="0 0 384 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s-12.5-32.8 0-45.3L237.3 256 342.6 150.6z"></path>
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div aria-live="polite" class="invalid-feedback" id="tagInput-validation">
          Required field.
        </div>
        <div aria-live="polite" class="error-message" id="tagInput-error">
          Invalid input
        </div>
      </div>
    </div>
  </div>
</autocomplete-multiselect>
`;
